File: backend/shopify_proxy.py
# shopify_proxy.py — Shopify App Proxy HMAC verification middleware.
#
# App-specific (not in Drink-X): the public events pages are served through the
# Shopify App Proxy at www.88bamboo.co/a/events/*. Shopify signs every proxied
# request with an HMAC-SHA256 `signature` query param computed from the other
# query params + the shared secret.
#
# Gating (plan §4/§9): verification runs ONLY when SHOPIFY_PROXY_VERIFY=true.
# Locally there is no proxy, so the flag is false and this middleware is a no-op.
# The shared secret is read from SHOPIFY_SHARED_SECRET (never hardcoded).

import hashlib
import hmac
import os

from flask import request, jsonify

# Paths that must never be gated by the proxy signature (health checks, etc.).
EXEMPT_PREFIXES = ("/health",)


def _truthy(value):
    return str(value).lower() in ("1", "true", "yes")


def verify_proxy_signature(args, secret):
    """Return True if `args` carry a valid Shopify App Proxy signature.

    Shopify builds the signature by taking every query param EXCEPT `signature`,
    sorting them by key, joining each as `key=value` (values joined with ',' when
    repeated) with no separator between pairs, then HMAC-SHA256 hex-digesting that
    string with the app's shared secret.
    """
    if not secret:
        return False

    provided = args.get("signature")
    if not provided:
        return False

    # Build the canonical string from all params except `signature`.
    params = {}
    for key in args.keys():
        if key == "signature":
            continue
        # request.args is a MultiDict; repeated values are comma-joined.
        params[key] = ",".join(args.getlist(key))

    message = "".join(f"{k}={params[k]}" for k in sorted(params.keys()))
    computed = hmac.new(
        secret.encode("utf-8"),
        message.encode("utf-8"),
        hashlib.sha256,
    ).hexdigest()

    return hmac.compare_digest(computed, provided)


def init_shopify_proxy(app):
    """Register the gated before_request hook on the Flask app."""

    @app.before_request
    def _verify_shopify_proxy():
        # Disabled unless explicitly turned on (false locally — no proxy exists).
        if not _truthy(os.getenv("SHOPIFY_PROXY_VERIFY", "false")):
            return None

        # Never gate health checks (load balancers / compose healthchecks).
        if request.path.startswith(EXEMPT_PREFIXES):
            return None

        secret = os.getenv("SHOPIFY_SHARED_SECRET")
        if not verify_proxy_signature(request.args, secret):
            return jsonify({"code": 401, "error": "Invalid App Proxy signature"}), 401

        return None

File: frontend/core/utils/shopifyProxy.js
// core/utils/shopifyProxy.js — Shopify App Proxy signature verification for the
// frontend (server-side only). Parity with the backend middleware
// (backend/shopify_proxy.py); gated by the same SHOPIFY_PROXY_VERIFY flag.
//
// Public event pages are fetched by Shopify (server-side) and delivered under
// www.88bamboo.co/a/events/*. Call verifyProxyRequest(ctx) inside a page's
// getServerSideProps to reject unsigned direct hits when the flag is on.
// Locally the flag is false (no proxy exists), so this is a no-op pass.
import crypto from 'crypto';

function isTruthy(value) {
  return String(value).toLowerCase() === 'true' ||
    String(value) === '1' ||
    String(value).toLowerCase() === 'yes';
}

/**
 * Verify a Shopify App Proxy signature from parsed query params.
 * Shopify sorts all params except `signature`, joins them as `key=value` with no
 * separator, and HMAC-SHA256 hex-digests that string with the shared secret.
 *
 * @param {Object} query - parsed query params (e.g. ctx.query)
 * @param {string} secret - SHOPIFY_SHARED_SECRET
 * @returns {boolean}
 */
export function verifyProxySignature(query, secret) {
  if (!secret) return false;
  const provided = query.signature;
  if (!provided) return false;

  const message = Object.keys(query)
    .filter((k) => k !== 'signature')
    .sort()
    .map((k) => `${k}=${Array.isArray(query[k]) ? query[k].join(',') : query[k]}`)
    .join('');

  const computed = crypto
    .createHmac('sha256', secret)
    .update(message)
    .digest('hex');

  try {
    return crypto.timingSafeEqual(Buffer.from(computed), Buffer.from(provided));
  } catch {
    return false; // length mismatch => invalid
  }
}

/**
 * Guard for getServerSideProps. Returns { valid: true } when verification is
 * disabled or the signature checks out; otherwise { valid: false }.
 *
 * On dynamic routes (e.g. pages/[slug].js), Next.js merges the route params
 * into ctx.query alongside the real query string. Shopify's signature is
 * only ever computed over the actual forwarded query string, so those
 * route-param keys (not real query params) must be stripped before
 * verifying, or every dynamic-route page fails signature verification.
 *
 * @param {import('next').GetServerSidePropsContext} ctx
 */
export function verifyProxyRequest(ctx) {
  if (!isTruthy(process.env.SHOPIFY_PROXY_VERIFY)) return { valid: true };
  const proxyQuery = { ...ctx.query };
  Object.keys(ctx.params || {}).forEach((key) => delete proxyQuery[key]);
  const ok = verifyProxySignature(proxyQuery, process.env.SHOPIFY_SHARED_SECRET);
  return { valid: ok };
}
